### PR TITLE
ME-75 - Send buyer id instead of buyer email for authenticated buyers

### PR DESCRIPTION
**Pull request by dhiraj-toptal**

### CODE REVIEW

File: Controller/Payment/Checkout/Token.php [hunk 1]
```diff
@@ -122,4 +141,28 @@ public function execute()
...
+     * @param $transactionId
+     */
+    public function createBuyer($transactionId)
+    {
+        try {
+            if ($this->customerSession->isLoggedIn() && empty($this->helper->getBuyerId()) && $transactionId) {
+                $response = $this->requestFactory
+                    ->create(RequestFactory::TRANSACTIONS_REQUEST_METHOD)
+                    ->setRequestMethod('transactions/' . $transactionId)
+                    ->setTopic('gettransactionid')
```
Comment (ynaxonBalance) [inline Controller/Payment/Checkout/Token.php]: What is this doing?

File: Controller/Payment/Checkout/Token.php [hunk 2]
```diff
@@ -122,4 +141,28 @@ public function execute()
...
+     *
+     * @param $transactionId
+     */
+    public function createBuyer($transactionId)
+    {
+        try {
+            if ($this->customerSession->isLoggedIn() && empty($this->helper->getBuyerId()) && $transactionId) {
+                $response = $this->requestFactory
+                    ->create(RequestFactory::TRANSACTIONS_REQUEST_METHOD)
+                    ->setRequestMethod('transactions/' . $transactionId)
```
Comment (ynaxonBalance) [inline Controller/Payment/Checkout/Token.php]: shouldn't request method be `POST`/`GET`/ etc?

File: Model/Config.php [hunk 3]
```diff
@@ -334,6 +334,19 @@ public function getAllowedPaymentMethods($scope = ScopeInterface::SCOPE_STORE, $
...
     }
 
+    /**
+     * Get AllowedTermsPaymentMethods
+     *
+     * @param string $scope
+     * @param int $storeId
+     * @return array|string[]
+     */
+    public function getAllowedTermsPaymentMethods($scope = ScopeInterface::SCOPE_STORE, $storeId = null)
```
Comment (ynaxonBalance) [inline Model/Config.php:344]: How is this related to this PR?

File: Model/Request/Transactions.php [hunk 4]
```diff
@@ -147,35 +211,44 @@ protected function getParams()
...
+        }
+        if ($quote->getCustomerIsGuest()) {
+            $params['email'] = $email;
+            $params['isRegistered'] = false;
+        } elseif ($this->customerSession->isLoggedIn() && empty($this->helper->getBuyerId())) {
+            $params['email'] = $email;
+            $params['isRegistered'] = false;
+        } elseif ($this->customerSession->isLoggedIn() && $this->helper->getBuyerId() != null) {
+            $params['id'] = $this->helper->getBuyerId($quote->getCustomerId());
+            $params['isRegistered'] = true;
```
Comment (ynaxonBalance) [inline Model/Request/Transactions.php]:
```
if ($this->customerSession->isLoggedIn() && $this->helper->getBuyerId() !== null) {
    $params['id'] = $this->helper->getBuyerId($quote->getCustomerId());
} else {
    $params['email'] = $email;
    $params['isRegistered'] = false;
}
```

Comment (ynaxonBalance) [reply Model/Request/Transactions.php]: (no need for `isRegistered` flag when passing `buyerId`

File: Model/Response/Transactions.php [hunk 5]
```diff
@@ -39,8 +44,9 @@ public function process()
         parent::process();
 
         $body = $this->getBody();
-        $this->_token = $body['token'];
+        $this->_token = isset($body['token']) ? $body['token'] : '';
```
Comment (ynaxonBalance) [inline Model/Response/Transactions.php:48]: is this the get transaction response? if so, token is always available

File: Model/Response/Transactions.php [hunk 6]
```diff
@@ -52,10 +58,7 @@ public function process()
      */
     protected function getRequiredResponseDataKeys()
     {
-        if ($this->_balancepayConfig->getIsAuth()) {
-            return ['token', 'id'];
-        }
-        return ['token'];
+        return [];
```
Comment (ynaxonBalance) [inline Model/Response/Transactions.php:62]: why did you make this change? both id and token always exist on get transaction response

File: Model/Response/Transactions.php [hunk 7]
```diff
@@ -77,4 +80,14 @@ public function getTransactionId()
     {
         return $this->_transactionId;
     }
+
+    /**
+     * GetTransactionId
```
Comment (ynaxonBalance) [inline Model/Response/Transactions.php]: This is get buyer id, not get transaction id

File: Controller/Payment/Checkout/Token.php [hunk 8]
```diff
@@ -122,4 +141,28 @@ public function execute()
...
+                    ->setTopic('gettransactionid')
+                    ->process();
+                if (!empty($response->getBuyerId())) {
+                    $this->helper->updateBuyerId($response->getBuyerId());
+                }
+            }
+        } catch (\Exception $e) {
```
Comment (ynaxonBalance) [inline Controller/Payment/Checkout/Token.php]: log and throw on error

File: Controller/Payment/Checkout/Token.php [hunk 9]
```diff
@@ -122,4 +141,28 @@ public function execute()
...
+
+    /**
+     * CreateBuyer
+     *
+     * @param $transactionId
+     */
+    public function createBuyer($transactionId)
+    {
+        try {
+            if ($this->customerSession->isLoggedIn() && empty($this->helper->getBuyerId()) && $transactionId) {
```
Comment (ynaxonBalance) [inline Controller/Payment/Checkout/Token.php]: move condition outside of this method – to keep method only responsible for creating the buyer

File: Controller/Payment/Checkout/Token.php [hunk 10]
```diff
@@ -93,12 +110,14 @@ public function execute()
...
             $result = $this->requestFactory
                 ->create(RequestFactory::TRANSACTIONS_REQUEST_METHOD)
+                ->setRequestMethod('transactions')
                 ->setFallbackEmail($fallbackEmail)
                 ->process();
-
             $token = $result->getToken();
             $transactionId = $result->getTransactionId();
 
+            $this->createBuyer($transactionId);
```
Comment (ynaxonBalance) [inline Controller/Payment/Checkout/Token.php]:
```
if (transactionId && $this->customerSession->isLoggedIn() && empty($this->helper->getBuyerId())) {
    $this->createBuyer($transactionId);
}

File: Controller/Payment/Checkout/Token.php [hunk 11]
```diff
@@ -122,4 +141,28 @@ public function execute()
             ->setHttpResponseCode(\Magento\Framework\Webapi\Response::HTTP_OK)
             ->setData($resBody);
     }
+
+    /**
+     * CreateBuyer
+     *
+     * @param $transactionId
+     */
+    public function createBuyer($transactionId)
```
Comment (ynaxonBalance) [inline Controller/Payment/Checkout/Token.php]: Let's move all buyer related functionality to more appropriate class (like `BalanceBuyer` class)

File: Helper/Data.php [hunk 12]
```diff
@@ -167,6 +184,37 @@ public function getBuyerAmount()
         return $response;
     }
 
+    /**
+     * Update Buyer Id
+     *
+     * @param $buyerId
+     */
+    public function updateBuyerId($buyerId)
```
Comment (ynaxonBalance) [inline Helper/Data.php]: Let's move all buyer related functionality to more appropriate class (like `BalanceBuyer` class)

File: etc/adminhtml/system.xml [hunk 13]
```diff
@@ -75,6 +75,12 @@
                     <source_model>Balancepay\Balancepay\Model\Adminhtml\Source\AllowedPaymentMethods</source_model>
                     <config_path>payment/balancepay/allowed_payment_methods</config_path>
                 </field>
+                <field id="net_terms_allowed_payment_methods" translate="label" type="multiselect" sortOrder="80" showInDefault="1" showInWebsite="1" showInStore="1" canRestore="1">
```
Comment (ynaxonBalance) [inline etc/adminhtml/system.xml:78]: How is this related to this PR?

File: Helper/Data.php (hunk 12 above)
Comment (ynaxonBalance) [reply Helper/Data.php]: consider renaming to `updateCustomerBalanceBuyerId`

File: Helper/Data.php [hunk 14]
```diff
@@ -167,6 +184,37 @@ public function getBuyerAmount()
...
+    }
+
+    /**
+     * GetBuyerId
+     *
+     * @return mixed|string
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function getBuyerId()
```
Comment (ynaxonBalance) [inline Helper/Data.php]: Let's move all buyer related functionality to more appropriate class (like BalanceBuyer class)

Comment (ynaxonBalance) [reply Helper/Data.php]: Consider renaming to `getCustomerBalanceBuyerId`

File: Model/Response/Transactions.php (hunk 5 above)
Comment (dhiraj-toptal) [reply Model/Response/Transactions.php:48]: This is called for both POST and GET method, so we have checked if token is set.

File: Model/Config.php (hunk 3 above)
Comment (dhiraj-toptal) [reply Model/Config.php:344]: This is related to ME-51, I forgot to remove related changes from this PR.

File: Controller/Payment/Checkout/Token.php (hunk 2 above)
Comment (dhiraj-toptal) [reply Controller/Payment/Checkout/Token.php]: For method they have used getCurlMethod, we are following same function further. 
We can make the change but have to make it for every request and need to test everything thoroughly.

File: Controller/Payment/Checkout/Token.php (hunk 1 above)
Comment (dhiraj-toptal) [reply Controller/Payment/Checkout/Token.php]: This calls the GET request for transaction with transactionID.

File: Model/Response/Transactions.php (hunk 5 above)
Comment (ynaxonBalance) [reply Model/Response/Transactions.php:48]: it will be set on both

Comment (dhiraj-toptal) [reply Model/Response/Transactions.php:48]: actually we don't get token from GET call response.

Comment (ynaxonBalance) [reply Model/Response/Transactions.php:48]: Why do we use the same logic for POST / GET?

Comment (dhiraj-toptal) [reply Model/Response/Transactions.php:48]: we have a Transactions.php which is a specific class which handles related requests and process function handles the responses for both these methods.

File: Model/Response/Transactions.php (hunk 6 above)
Comment (dhiraj-toptal) [reply Model/Response/Transactions.php:62]: Token doesn't exists on the GET response.

File: etc/adminhtml/system.xml (hunk 13 above)
Comment (dhiraj-toptal) [reply etc/adminhtml/system.xml:78]: This is related to ME-51, I forgot to remove related changes from this PR.

**Comment by dhiraj-toptal** [comment]
Hi @ynaxonBalance 
I have updated PR. Can you please review?

File: Model/BalanceBuyer.php [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            $response = $this->requestFactory
+                ->create(RequestFactory::TRANSACTIONS_REQUEST_METHOD)
+                ->setRequestMethod('transactions/' . $transactionId)
+                ->setTopic('gettransactionid')
+                ->process();
+            if (!empty($response->getBuyerId())) {
+                $this->updateCustomerBalanceBuyerId($response->getBuyerId());
+            }
+        } catch (\Exception $e) {
+            $this->logger->error('There is an while fetching Buyer Id from transaction.');
```
Comment (ynaxonBalance) [inline Model/BalanceBuyer.php]: log transactionId and error object

Comment (ynaxonBalance) [reply Model/BalanceBuyer.php]: Error message:

> Could not attach buyer id to the customer

Comment (ynaxonBalance) [reply Model/BalanceBuyer.php]:
```
[
    'transactionId' => '...',
    'httpError' => [
          error => '...',
          message =>  '...',
          statusCode =>  400
     ]
]
```

File: Model/Request/Transactions.php (hunk 4 above)
Comment (dhiraj-toptal) [reply Model/Request/Transactions.php]: I have updated above conditions

File: Model/BalanceBuyer.php (hunk 15 above)
Comment (dhiraj-toptal) [reply Model/BalanceBuyer.php]: I have added more data to the log.

**Review by ynaxonBalance — APPROVED** [review]
